src/core/model/reducer.ts

```diff
@@ -49,10 +58,34 @@ export const uiSchemaReducer = (
     case ADD_UNSCOPED_ELEMENT_TO_LAYOUT:
       return withCloneTree(action.layout, uiSchema, (newUiSchema) => {
         const newUIElement = action.uiSchemaElement;
-        (newUIElement as LinkedUISchemaElement).parent = newUiSchema;
+        newUIElement.parent = newUiSchema;
         newUiSchema.elements.splice(action.index, 0, newUIElement);
         return getRoot(newUiSchema as LinkedUISchemaElement);
       });
+    case MOVE_UISCHEMA_ELEMENT:
+      return withCloneTree(
+        action.layout as LinkedUISchemaElement,
+        uiSchema,
+        (newUiSchema) => {
+          const elementToMove = action.uiSchemaElement as LinkedUISchemaElement;
+
+          //remove element from old parent
+          const updatedUiSchema = removeUiElement(elementToMove, newUiSchema);
+          if (isRemoveElementError(updatedUiSchema) || !updatedUiSchema) {
+            return uiSchema;
+          }
+          //add element to new parent
+          elementToMove.parent = updatedUiSchema;
+          const index = action.index;
+          (updatedUiSchema as Layout).elements.splice(
+            index - 1,
```

Comment:
I think the `-1` is wrong, all indexes should be the same, so we don't need a 1-based to 0-based conversion. This also explains the error that if you move an element below another, it gets added above, and adding it to the first position behaves weird, it works for horizontal but not for vertical.

Reply:
The  -1 is here because we don't want to allow dragging the element in the drop point right after the element, which will result in the same layout.